cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/ZookeeperScaler.java

```diff
@@ -0,0 +1,351 @@
+/*
+ * Copyright Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.operator.cluster.operator.resource;
+
+import io.fabric8.kubernetes.api.model.Secret;
+import io.strimzi.operator.cluster.model.Ca;
+import io.strimzi.operator.cluster.model.ZookeeperCluster;
+import io.strimzi.operator.common.Util;
+import io.vertx.core.Future;
+import io.vertx.core.Promise;
+import io.vertx.core.Vertx;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.zookeeper.KeeperException;
+import org.apache.zookeeper.admin.ZooKeeperAdmin;
+import org.apache.zookeeper.client.ZKClientConfig;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Class for scaling Zookeeper 3.5 using the ZookeeperAdmin client
+ */
+public class ZookeeperScaler implements AutoCloseable {
+    private static final Logger log = LogManager.getLogger(ZookeeperScaler.class);
+
+    private final Vertx vertx;
+    private final ZooKeeperAdminProvider zooAdminProvider;
+
+    private final String zookeeperConnectionString;
+
+    private final long operationTimeoutMs;
+
+    private final String trustStorePassword;
+    private final File trustStoreFile;
+
+    private final String keyStorePassword;
+    private final File keyStoreFile;
+
+    private ZooKeeperAdmin zkAdmin;
```

Comment:
This is accessed in various places by callbacks to `vertx.executeBlocking()`, i.e. potentially by different threads. Probably the correct thing to do is for `connect()` to return `Future<ZooKeeperAdmin>` and propagate the client through the compose chain in `scale()`, though that's a bit messy because `getConfig()` would need to return both the config and the client instance.

Reply:
Yeah ... I was a bit strugling with how put it together. It seemed to be best this way also because of closing the connections in case of failures etc. Java makes this hard with not having any Tuples etc. We would probably need some custom class to pass both the results and the client along. The ZooKeeperAdmin client is also a bit weird mixture of synchronous and asynchronous methods which makes it harder to structure it. Connection is async, getConfig or reconfigure are sync.

Reply:
Yeah, this was one of the issues which led me to ultimately change the `KafkaRoller` to be single threaded. I think a custom class would be fine, it only needs two `final` fields and a constructor (roll on Java 14 with records).